From 7feb784dc2c6f5a5e8aacbc9672d87ebacbe42dc Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 12 Sep 2024 17:17:57 +0200
Subject: [PATCH 1203/1695] plat-stm32mp2: add suspend/resume callback for HSE
 monitoring

Stop the counter at suspend and re-start it at the resume after a low
power mode, when HSE is stopped.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git af7d401c6861661898a1f01c7b95bd99c4947e54 plat-stm32mp2: add suspend/resume callback for HSE monitoring]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I496962177a9d1b3372dd280d528c477717288450
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/404189
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c | 35 +++++++++++++++++++++++++++++-
 1 file changed, 34 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index 122b15c2c..c9e3812fc 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -26,6 +26,7 @@
 #include <kernel/dt.h>
 #include <kernel/interrupt.h>
 #include <kernel/misc.h>
+#include <kernel/pm.h>
 #include <kernel/spinlock.h>
 #include <mm/core_memprot.h>
 #include <platform_config.h>
@@ -380,6 +381,11 @@ TEE_Result stm32_activate_internal_tamper(int id)
 #define HSE_ALARM_PERCENT	110
 #define FREQ_MONITOR_COMPAT	"st,freq-monitor"
 
+struct stm32_hse_monitoring_data {
+	struct counter_device *counter;
+	void *config;
+};
+
 static void stm32_hse_over_frequency(uint32_t ticks __unused,
 				     void *user_data __unused)
 {
@@ -387,8 +393,27 @@ static void stm32_hse_over_frequency(uint32_t ticks __unused,
 }
 DECLARE_KEEP_PAGER(stm32_hse_over_frequency);
 
+static TEE_Result stm32_hse_monitoring_pm(enum pm_op op,
+					  unsigned int pm_hint __unused,
+					  const struct pm_callback_handle *h)
+{
+	struct stm32_hse_monitoring_data *priv =
+	(struct stm32_hse_monitoring_data *)PM_CALLBACK_GET_HANDLE(h);
+
+	if (op == PM_OP_RESUME) {
+		counter_start(priv->counter, priv->config);
+		counter_set_alarm(priv->counter);
+	} else {
+		counter_cancel_alarm(priv->counter);
+		counter_stop(priv->counter);
+	}
+
+	return TEE_SUCCESS;
+}
+
 static TEE_Result stm32_hse_monitoring(void)
 {
+	struct stm32_hse_monitoring_data *priv = NULL;
 	struct counter_device *counter = NULL;
 	struct clk *hse_clk = NULL;
 	struct clk *hsi_clk = NULL;
@@ -400,7 +425,9 @@ static TEE_Result stm32_hse_monitoring(void)
 	int node = 0;
 	int res  = 0;
 
-	DMSG("HSE monitoring");
+	priv = calloc(1, sizeof(*priv));
+	if (!priv)
+		return TEE_ERROR_OUT_OF_MEMORY;
 
 	fdt = get_embedded_dt();
 	node = fdt_node_offset_by_compatible(fdt, 0, FREQ_MONITOR_COMPAT);
@@ -438,6 +465,12 @@ static TEE_Result stm32_hse_monitoring(void)
 	counter->alarm.callback = stm32_hse_over_frequency;
 	counter->alarm.ticks = ticks;
 
+	priv->counter = counter;
+	priv->config = config;
+
+	register_pm_core_service_cb(stm32_hse_monitoring_pm, priv,
+				    "stm32-hse-monitoring");
+
 	counter_start(counter, config);
 	counter_set_alarm(counter);
 
-- 
2.47.2

